src/player.cpp

```diff
@@ -5262,14 +5252,33 @@ void player::process_active_items()
             active_held_items.push_back( index );
         }
     }
-    // Necessary for UPS in Aftershock - check worn items for charge
-    for( const item &it : worn ) {
-        itype_id identifier = it.type->get_id();
+    bool update_required = get_check_encumbrance();
+    for( item &w : worn ) {
+        if( w.has_flag( "USE_UPS" ) &&
+            w.charges < w.type->maximum_charges() ) {
+            active_worn_items.push_back( &w );
+        }
+        if( cloak == nullptr && w.has_flag( "ACTIVE_CLOAKING" ) ) {
+            cloak = &w;
+        }
+        // Only the main power armor item can be active, the other ones (hauling frame, helmet) aren't.
+        if( power_armor == nullptr && w.is_power_armor() ) {
+            power_armor = &w;
+        }
+        // Necessary for UPS in Aftershock - check worn items for charge
+        itype_id identifier = w.type->get_id();
```

Comment:
```suggestion
        const itype_id &identifier = w.typeId();
```
Avoiding the copy is also a performance boost.